src/librustdoc/passes/calculate_doc_coverage.rs

```diff
@@ -64,13 +67,73 @@ impl ops::AddAssign for ItemCount {
     }
 }
 
-#[derive(Default)]
 struct CoverageCalculator {
     items: BTreeMap<FileName, ItemCount>,
+    output_format: Option<OutputFormat>,
+}
+
+fn limit_filename_len(filename: String) -> String {
+    let nb_chars = filename.chars().count();
+    if nb_chars > 35 {
+        "...".to_string()
+            + &filename[filename.char_indices().nth(nb_chars - 32).map(|x| x.0).unwrap_or(0)..]
+    } else {
+        filename
+    }
+}
+
+#[derive(Serialize)]
+struct CoverageCalculatorEntry {
+    documented: u64,
+    total: u64,
+    percentage: f64,
 }
 
 impl CoverageCalculator {
+    fn new(output_format: Option<OutputFormat>) -> CoverageCalculator {
+        CoverageCalculator { items: Default::default(), output_format }
+    }
+
+    fn to_json(&self) -> String {
+        let mut total = ItemCount::default();
+
+        let mut json = self
+            .items
+            .iter()
+            .filter_map(|(file, &count)| {
+                count
+                    .percentage()
+                    .map(|percent| (limit_filename_len(file.to_string()), count, percent))
+            })
+            .map(|(name, count, percentage)| {
+                total += count;
+
+                (
+                    name,
+                    CoverageCalculatorEntry {
+                        documented: count.with_docs,
+                        total: count.total,
+                        percentage: percentage,
+                    },
+                )
+            })
+            .collect::<BTreeMap<String, CoverageCalculatorEntry>>();
+        json.insert(
+            "total".to_owned(),
+            CoverageCalculatorEntry {
+                documented: total.with_docs,
+                total: total.total,
+                percentage: total.percentage().unwrap_or(0.0),
+            },
+        );
+        serde_json::to_string_pretty(&json).expect("failed to convert JSON data to string")
```

Comment:
If the JSON is to be consumed by some other tool then it doesn't need to be pretty printed.